src/lib/mf.rs

```diff
@@ -509,16 +476,398 @@ fn rank_cnds<TokId: Clone + Copy + Debug + TryFrom<usize> + TryInto<usize> + Par
+        // might end up in.
+        //
+        // We calculate this for a state i, production p_idx, by iterating backwards over the
+        // stategraph (using rev_edges) finding all routes of length grm.prod(p_idx).len() which
+        // could lead back to i, and then storing their goto state indexes. Since further
+        // reductions and gotos could then happen, we then have to recursively search those goto
+        // state indexes from scratch.
+        //
+        // This loop is currently comically inefficient, but it is relatively straightforward.
+        let mut goto_edges = HashMap::new();
+        for i in 0..states_len {
+            for &(p_idx, sym_off) in sgraph.core_state(StIdx::from(i)).items.keys() {
+                let prod = grm.prod(p_idx);
+                if usize::from(sym_off) == prod.len() {
+                    let mut todo = HashSet::new(); // (StIdx, PIdx)
+                    let mut done = HashSet::new(); // Which (StIdx, PIdx) pairs have we processed?
+                    let mut ge = HashSet::new();   // Which StIdx's can we eventually goto?
+                    todo.insert((i, p_idx));
+                    while todo.len() > 0 {
+                        let &(i, p) = todo.iter().nth(0).unwrap();
+                        todo.remove(&(i, p));
+                        if done.contains(&(i, p)) {
+                            continue;
+                        }
+                        done.insert((i, p));
+                        let mut cur = rev_edges[i].clone();
+                        let mut next;
+                        for _ in 0..grm.prod(p).len() - 1 {
+                            next = HashSet::new();
+                            for st_idx in cur {
+                                next.extend(&rev_edges[usize::from(st_idx)]);
+                            }
+                            cur = next;
+                        }
+                        for st_idx in cur {
+                            for &(sub_p_idx, sub_sym_off) in sgraph.core_state(StIdx::from(st_idx))
+                                                                   .items
+                                                                   .keys() {
+                                let sub_prod = grm.prod(sub_p_idx);
+                                if usize::from(sub_sym_off) == sub_prod.len() {
+                                    todo.insert((usize::from(st_idx), sub_p_idx));
+                                }
+                            }
+                            let n = grm.prod_to_nonterm(p);
+                            if let Some(goto_idx) = stable.goto(st_idx, n) {
+                                ge.insert(goto_idx);
+                            }
+                        }
+                    }
+                    goto_edges.insert((i, p_idx), ge);
+                }
+            }
+        }
+
+        // With goto_edges calculated, we can now percolate its effects through the dist table.
```

Comment:
percolate = filter?

Reply:
They're not quite synonyms, though I agree they're very similar. But I agree with your implicit criticism that "percolate" isn't as widely known a word. The comment is also a little out of date (long story), so I think I can rephrase and improve.